hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestIncrementalReadWithFullTableScan.scala

```diff
@@ -0,0 +1,187 @@
+
+    // Test both start commit and end commits are not archived but got cleaned
+    shouldThrowIfFallbackIsFalse(tableType,
+      () => runIncrementalQueryAndCompare(startUnarchivedCommitTs, endUnarchivedCommitTs, 1, false))
+    runIncrementalQueryAndCompare(startUnarchivedCommitTs, endUnarchivedCommitTs, 1, true)
+
+    // Test start commit is not archived, end commits is out of the timeline
+    runIncrementalQueryAndCompare(startUnarchivedCommitTs, endOutOfRangeCommitTs, 5, true)
+
+    // Test both start commit and end commits are out of the timeline
+    runIncrementalQueryAndCompare(startOutOfRangeCommitTs, endOutOfRangeCommitTs, 0, false)
+    runIncrementalQueryAndCompare(startOutOfRangeCommitTs, endOutOfRangeCommitTs, 0, true)
+
+    // Test end commit is smaller than the start commit
+    runIncrementalQueryAndCompare(endUnarchivedCommitTs, startUnarchivedCommitTs, 0, false)
+    runIncrementalQueryAndCompare(endUnarchivedCommitTs, startUnarchivedCommitTs, 0, true)
+
+    // Test both start commit and end commits is not archived and not cleaned
+    val reversedCommits = completedCommits.getReverseOrderedInstants.toArray
+    val startUncleanedCommitTs = reversedCommits.apply(1).asInstanceOf[HoodieInstant].getTimestamp
+    val endUncleanedCommitTs = reversedCommits.apply(0).asInstanceOf[HoodieInstant].getTimestamp
+    runIncrementalQueryAndCompare(startUncleanedCommitTs, endUncleanedCommitTs, 1, true)
+    runIncrementalQueryAndCompare(startUncleanedCommitTs, endUncleanedCommitTs, 1, false)
+  }
+
+  private def runIncrementalQueryAndCompare(
+      startTs: String,
+      endTs: String,
+      batchNum: Int,
+      fallBackFullTableScan: Boolean): Unit = {
+    val hoodieIncViewDF = spark.read.format("org.apache.hudi")
+      .option(DataSourceReadOptions.QUERY_TYPE.key(), DataSourceReadOptions.QUERY_TYPE_INCREMENTAL_OPT_VAL)
+      .option(DataSourceReadOptions.BEGIN_INSTANTTIME.key(), startTs)
+      .option(DataSourceReadOptions.END_INSTANTTIME.key(), endTs)
+      .option(DataSourceReadOptions.INCREMENTAL_FALLBACK_TO_FULL_TABLE_SCAN_FOR_NON_EXISTING_FILES.key(), fallBackFullTableScan)
+      .load(basePath)
+    assertEquals(perBatchSize * batchNum, hoodieIncViewDF.count())
+  }
+
+  private def shouldThrowIfFallbackIsFalse(tableType: HoodieTableType, fn: () => Unit): Unit = {
+    val msg = "Should fail with Path does not exist"
+    tableType match {
+      case HoodieTableType.COPY_ON_WRITE =>
+        assertThrows(classOf[AnalysisException], new Executable {
+          override def execute(): Unit = {
+            fn()
+          }
+        }, msg)
+      case HoodieTableType.MERGE_ON_READ =>
+        val exp = assertThrows(classOf[SparkException], new Executable {
+          override def execute(): Unit = {
+            fn()
+          }
+        }, msg)
+        assertTrue(exp.getMessage.contains("FileNotFoundException"))
+    }
+  }
```

Comment:
Can we just execute the code block first:
```java
        assertThrows(classOf[AnalysisException], new Executable {
          override def execute(): Unit = {
            fn()
          }
        }, msg)
```

Then check the exception msg especially for MERGE_ON_READ table ?

Reply:
It will throw two different exception type for MOR tables and COW tables,

for COW table it throws `AnalysisException`
for MOR table it throws `SparkException`